### PR TITLE
[Maybe breaking change] Fix day of year computation

### PR DESCRIPTION
**Pull request by ChristopherRabotin**
The day of year computation was broken: first, it didn't work well when the time scale reference epoch was not J1900 (cf #272 ), and moreover, it started counting the days at zero, when it should have one (to meet the GPS standard and the Python behavior).

### CODE REVIEW

File: src/duration.rs [hunk 1]
```diff
@@ -1008,6 +1008,7 @@ impl Add for Duration {
     /// ## Examples
     /// + `Duration { centuries: 0, nanoseconds: 1 }` is a positive duration of zero centuries and one nanosecond.
     /// + `Duration { centuries: -1, nanoseconds: 1 }` is a negative duration representing "one century before zero minus one nanosecond"
+    #[allow(clippy::absurd_extreme_comparisons)]
```
Comment (ChristopherRabotin) [inline src/duration.rs:1011]: Prevent what clippy said was useless but actually required.

File: src/efmt/format.rs [hunk 2]
```diff
@@ -90,7 +90,7 @@ const MAX_TOKENS: usize = 16;
 /// assert_eq!(fmt, consts::ISO8601_ORDINAL);
 ///
 /// let fmt_iso_ord = Formatter::new(bday, consts::ISO8601_ORDINAL);
-/// assert_eq!(format!("{fmt_iso_ord}"), "2000-059");
+/// assert_eq!(format!("{fmt_iso_ord}"), "2000-060");
```
Comment (ChristopherRabotin) [inline src/efmt/format.rs:93]: Behavior change introduced by this PR.

File: src/efmt/format.rs [hunk 3]
```diff
@@ -546,10 +546,18 @@ fn epoch_format_from_str() {
...
+    >>> from datetime import datetime
+    >>> dt, fmt = "2023-117T12:55:26", "%Y-%jT%H:%M:%S"
+    >>> datetime.strptime(dt, fmt)
+    datetime.datetime(2023, 4, 27, 12, 55, 26)
+     */
+
     let e = Epoch::from_format_str("2023-117T12:55:26", "%Y-%jT%H:%M:%S").unwrap();
 
-    assert_eq!(format!("{e}"), "2023-04-28T12:55:26 UTC");
+    assert_eq!(format!("{e}"), "2023-04-27T12:55:26 UTC");
```
Comment (ChristopherRabotin) [inline src/efmt/format.rs:562]: Now matches Python as per the comment.

**Comment by ChristopherRabotin** [comment]
This should be tagged at 3.9.0 because of that behavior change.

**Review by gwbres — APPROVED** [review]

**Comment by gwbres** [comment]
I cannot push contributions to this PR since I did not initiate it.
Though  I did check that the test you wrote (for GPST + UTC) does pass in other timescales.
I may provide these additional tests in a separate PR.

**Comment by ChristopherRabotin** [comment]
Ok. If you paste the tests here, I can add them to the code base.

**Comment by gwbres** [comment]
Ok,

here's a test that does pass

```rust
#[test]
fn regression_test_gh_272() {
    use core::str::FromStr;

    let epoch = Epoch::from_str("2021-12-21T00:00:00 GPST").unwrap();

    let (years, day_of_year) = epoch.year_days_of_year();

    assert!(dbg!(day_of_year) < DAYS_PER_YEAR);
    assert!(day_of_year > 0.0);
    assert_eq!(day_of_year, 355.0);

    assert_eq!(years, 2021);

    // Check that we start counting the days at one, in all timescales.
    for ts in ["TAI", "UTC", "GPST", "GST", "BDT"] {
        let epoch = Epoch::from_str(&format!("2021-12-31T00:00:00 {}", ts)).unwrap();
        let (years, day_of_year) = epoch.year_days_of_year();
        assert_eq!(years, 2021);
        assert_eq!(day_of_year, 365.0);

        let epoch = Epoch::from_str(&format!("2020-12-31T00:00:00 {}", ts)).unwrap();
        let (years, day_of_year) = epoch.year_days_of_year();
        assert_eq!(years, 2020);
        // 366 days in 2020, leap year.
        assert_eq!(day_of_year, 366.0);

        let epoch = Epoch::from_str(&format!("2021-01-01T00:00:00 {}", ts)).unwrap();
        let (years, day_of_year) = epoch.year_days_of_year();
        assert_eq!(years, 2021);
        assert_eq!(day_of_year, 1.0);
    }
}

```

**Comment by ChristopherRabotin** [comment]
I added those. The only difference is that I avoided the `format!()` instruction because it's only available with the `std` feature.

**Comment by ChristopherRabotin** [comment]
@gwbres Let me know if this looks good to merge and would fix your usecase.

**Comment by gwbres** [comment]
Hello, yes it alll looks fine to me

**Pull request merged**